lib/maths/unittest/CSamplingTest.cc

```diff
@@ -214,4 +219,104 @@ BOOST_AUTO_TEST_CASE(testMultivariateNormalSample) {
     }
 }
 
+BOOST_AUTO_TEST_CASE(testReservoirSampling) {
+
+    // Check we uniformly sample from a stream.
+
+    using TSampler = maths::CSampling::CReservoirSampler<double>;
+    using TMeanAccumulator = maths::CBasicStatistics::SSampleMean<double>::TAccumulator;
+
+    TDoubleVec samples(200);
+    TSampler sampler{200, [&](std::size_t slot, const double& value) {
+                         samples[slot] = value;
+                     }};
+
+    TMeanAccumulator pValue;
+    for (std::size_t t = 0; t < 100; ++t) {
+        samples.assign(200, 0.0);
+        sampler.reset();
+        for (double x = 0.0; x < 1000.0; x += 1.0) {
+            sampler.sample(x);
+        }
+
+        maths::CStatisticalTests::CCramerVonMises cvm{20};
+        for (const auto& sample : samples) {
+            cvm.addF(sample / 1000.0);
+        }
+
+        // The p-value is small if the samples *aren't* distributed as expected.
+        BOOST_TEST_REQUIRE(cvm.pValue() > 0.05);
+        pValue.add(cvm.pValue());
+    }
+
+    LOG_DEBUG(<< "mean p-value = " << maths::CBasicStatistics::mean(pValue));
+    BOOST_TEST_REQUIRE(maths::CBasicStatistics::mean(pValue) > 0.3);
+}
+
+BOOST_AUTO_TEST_CASE(testVectorDissimilaritySampler) {
+
+    // Test the average distance between points is significantly larger than
+    // for uniform random sampling.
+
+    using TVector = maths::CDenseVector<double>;
+    using TVectorVec = std::vector<TVector>;
+    using TReservoirSampler = maths::CSampling::CReservoirSampler<TVector>;
+    using TDissimilaritySampler = maths::CSampling::CVectorDissimilaritySampler<TVector>;
+    using TMeanAccumulator = maths::CBasicStatistics::SSampleMean<double>::TAccumulator;
+
+    std::size_t numberSamples{100};
+
+    TVectorVec samples(numberSamples);
+    TReservoirSampler randomSampler{
+        numberSamples,
+        [&](std::size_t slot, const TVector& value) { samples[slot] = value; }};
+    TDissimilaritySampler dissimilaritySampler{numberSamples};
+
+    test::CRandomNumbers rng;
+
+    TMeanAccumulator percentageSeparationIncrease;
+
+    TVector x{4};
+    TDoubleVec p;
+    TDoubleVec components;
```

Comment:
Should you reserve the space for the vectors accordingly?

Reply:
`rng.generateLogNormalSamples` calls reserve for me below. However, `p` is a hangover from an earlier version so I'll tidy up.

Reply:
[9c43087](https://github.com/elastic/ml-cpp/pull/1082/commits/9c43087df595c1eb012d76465343487470fac424).